corehq/apps/locations/models.py

```diff
@@ -430,22 +430,57 @@ def save(self, *args, **kwargs):
         publish_location_saved(self.domain, self.location_id)
 
     def delete(self, *args, **kwargs):
-        from corehq.apps.commtrack.models import sync_supply_point
+        """Delete this location and all descentants
+
+        Supply point cases and user updates are performed asynchronously.
+        """
+        from .tasks import update_users_at_locations
         from .document_store import publish_location_saved
-        to_delete = self.get_descendants(include_self=True)
 
-        # This deletion should ideally happen in a transaction. It's not
-        # currently possible as supply point cases are stored either in a
-        # separate database or in couch. Happy Debugging!
+        to_delete = self.get_descendants(include_self=True)
         for loc in to_delete:
-            loc._remove_users()
-            sync_supply_point(loc, is_deletion=True)
+            loc._remove_user()
 
         super(SQLLocation, self).delete(*args, **kwargs)
+        update_users_at_locations.delay(
+            self.domain,
+            [loc.location_id for loc in to_delete],
+            [loc.supply_point_id for loc in to_delete if loc.supply_point_id],
+            list(self.get_ancestors().location_ids()),
+        )
         publish_location_saved(self.domain, self.location_id, is_deletion=True)
 
     full_delete = delete
 
+    @classmethod
+    def bulk_delete(cls, locations, ancestor_location_ids):
```

Comment:
This might make more sense as a utility in bulk_management - it operates on the expectation that `locations` is a sanitized data set, where you wouldn't, for instance pass in a location and not also its children.  That validation happens elsewhere in bulk management, but if it's to be a general utility, I think it should cascade deletes.